### PR TITLE
Contest Framework: Support to download the test logs

### PR DESCRIPTION
**Pull request by susmithp**
Added the support for downloading the contest test log files.

### CODE REVIEW

File: misc/contest/contestcli.go [hunk 1]
```diff
@@ -24,30 +26,39 @@ func main() {
 		fmt.Println("Invalid inputs")
 		return
 	}
-	logFile := path.Join(*logpath, "contest_"+*user+".log")
+	timestamp := time.Now().Format("20060102150405")
+	logFile := path.Join(*logpath, "contest_"+*user+"_"+timestamp+".log")
 	loghandler, err := os.OpenFile(logFile, os.O_CREATE|os.O_APPEND|os.O_RDWR, 0666)
```
Comment (arunkoshy) [inline misc/contest/contestcli.go:31]: @susmithp let's get this reworked based on discussion offline.

**Review by arunkoshy — CHANGES_REQUESTED** [review]

File: misc/contest/contestcli.go [hunk 2]
```diff
@@ -24,30 +26,39 @@ func main() {
...
+	log.Printf("User:%s, Tests:%s\n", user, testlist)
 	for _, test := range tests {
 		if _, err := os.Stat(test); os.IsNotExist(err) {
 			log.Printf("File [%s] does not exist\n", test)
 			continue
 		}
 		log.Printf("Executing the test: %s\n", test)
+		testname := getTestName(test)
+		outfile := path.Join(userDir, "contest_"+testname+"_output.log")
+		outhandler, _ := os.OpenFile(outfile, os.O_CREATE|os.O_APPEND|os.O_RDWR, 0666)
```
Comment (arunkoshy) [inline misc/contest/contestcli.go:60]: @susmithp ditto to prior check -- let's discuss for this and ctrl1 situation

**Review by arunkoshy — CHANGES_REQUESTED** [review]

**Review by arunkoshy — APPROVED** [review]
LGTM @vejmarie

File: gateway/js/contest.js [hunk 3]
```diff
@@ -50,7 +50,36 @@ $(document).ready(function() {
...
+                                   console.log(percentage)
+                           }
+                        },
+                        success: function(response){
+                                $("#modalDownload").modal("hide");
+                                $("#modalDownloadBody").html("Download completed");
+                                var link=document.createElement('a');
+                                var url = window.URL || window.webkitURL;
+                                link.href=url.createObjectURL(response);
+                                link.download="contest_log.zip";
```
Comment (arunkoshy) [inline gateway/js/contest.js:79]: @susmithp we will refactor this to reflect the date/time stamp bits from the server location.

**Review by arunkoshy — APPROVED** [review]

**Review by vejmarie — APPROVED** [review]

File: ctrl/ctrl1.go [hunk 4]
```diff
@@ -629,6 +630,28 @@ func home(w http.ResponseWriter, r *http.Request) {
...
 		}
 		conn.Close()
+	case "test_logs":
+		_, tail := ShiftPath(r.URL.Path)
+		path := strings.Split(tail, "/")
+		username := path[1]
+		base.Zlog.Infof("Fetching the Test logs for user:%s", username)
+		command := "ls -tp " + testLog + "/contest_" + username + "*.zip | head -1"
+		base.Zlog.Infof("Executing command: %s", command)
+		cmd := exec.Command("/bin/sh", "-c", command)
```
Comment (arunkoshy) [inline ctrl/ctrl1.go:640]: @susmithp candidate for refactoring as per discussion offline

**Pull request merged**